src/Host/Client/Impl/Session/RSession.cs

```diff
@@ -262,34 +266,31 @@ internal sealed class RSession : IRSession, IRCallbacks {
 
             await _initializationTcs.Task;
             initializationCts.Dispose();
+            _stopHostLock.EnqueueReset();
             lockToken.Set();
         }
 
-        public IRSessionSwitchBrokerTransaction StartSwitchingBroker() {
-            _disposeToken.ThrowIfDisposed();
-            return new BrokerTransaction(this);
-        }
+        public IRSessionSwitchBrokerTransaction StartSwitchingBroker() => !_disposeToken.IsDisposed ? new BrokerTransaction(this) : null;
 
-        public IRSessionReconnectTransaction StartReconnecting() {
-            _disposeToken.ThrowIfDisposed();
-            return new BrokerTransaction(this);
-        }
+        public IRSessionReconnectTransaction StartReconnecting() => !_disposeToken.IsDisposed ? new BrokerTransaction(this) : null;
 
         public async Task StopHostAsync() {
             _disposeToken.ThrowIfDisposed();
```

Comment:
We can't guarantee that cancellation is completed, because we have no idea how many tasks are actually subscribed for that cancellation token and when do they check it. We also can't control the scheduler that may postpone one task in favor of another. But what we can be sure is that all tasks have received notification before `RequestCancellation` is completed, so the next time any code in the task will check for it, it will get an exception.

Ok then, I'll proceed with this option, and add this feature to `DisposableToken` and `DisposableBag`.

---
In reply to: [83955367](https://github.com/Microsoft/RTVS/pull/2529#discussion_r83955367) [](ancestors = 83955367)